docs/design/design.tex

```diff
@@ -1878,77 +1915,180 @@ \subsubsection{I2C host}\label{s:cl-i2c}
-    I2C_TOKEN_ADDR_READC = 0x4,
-    /* STOP: Used to send the STOP condition on the bus to end a transaction.
-    Causes master to release the bus. */
-    I2C_TOKEN_STOP = 0x5,
-  };
+  label={l:i2ccmddef},
+  caption={I2C command structure.}]
+  typedef struct i2c_cmd {
+    union {
+        /* Location of read destination / write data in data region */
+        uint8_t *data;
+
+        /* Fields for header members */
+        struct {
+            uint16_t batch_len;
+            i2c_addr_t address;
+        } i2c_header;
+    } payload;
+    /* Size of read/write operation. Max 256 back-to-back reads / 64KiB write */
+    /* Unused if this is a header command */
+    uint8_t data_len;
+    i2c_cmd_flags_t flag_mask;
+  } i2c_cmd_t;
 \end{lstlisting}
 
-Note: a single I2C transaction can only be for one address.
+It is an invariant that a header specifying a transaction of N commands must be followed
+by those N commands immediately. It is not possible to enqueue partial requests as this
+will significantly complicate the design of the device class.
 
-The Driver will parse the request and potentially divide it into smaller segments,
-each of which will be loaded into \gls{mmio} device registers
-on an interface of an explicitly, initially configured \gls{i2c} bus line.
-If the request is a READ, for each segment of the request a corresponding response
-will be loaded into the \gls{mmio} device registers by the device.
-The driver uses this data to build a response which re-uses the request buffer.
-Once the request has been completely loaded through the registers and converted into a response,
-the resulting response buffer will be enqueued by the Driver into the \gls{rs} queue.
-This response now travels only along the \gls{rs} interfaces between Driver, Virtualiser and Client.
+The response queue is a simple first-in, first-out queue containing simple response
+structures. Each response contains the I2C bus address of the corresponding request,
+the error value and the number of bytes written or read.
 
-\autoref{l:i2c_queue_entry} shows the format of the entries in the \gls{rq}
-and \gls{rs} queues.
+% TODO: diagram of a chain of commands with their header, showing buffers pointing to shared data region.
 
-\begin{lstlisting}[gobble=2,firstline=2,float=th,
-  label={l:i2c_queue_entry},
-  caption={I2C request and response queue entry data structure.}]
+\subsubsection{Clients and libi2c}
 
-  struct i2c_queue_entry {
-    size_t offset;
-    unsigned int len;
-    size_t bus_address;
-  };
-\end{lstlisting}
+Clients can request I2C transactions by composing a chain of requests locally and enqueuing
+them to the virtualiser queue. The virtualiser will process the contents of the queue upon
+being notified by the client. The client can manually compose an arbitrary transaction by
+creating commands or it can use libi2c to simplify transactions.
+
+\texttt{libi2c} is a simple blocking interface to the \gls{i2c} driver subsystem, supplying
```

Comment:
I wouldn't talk about libi2c in the design doc, that's more for developers.

You can just say we have blocking and non-blocking variants of the API for clients.

Reply:
Sure, will remove. Re: code and pseudocode, we probably don't need any. I only added what's there to replace what was removed.